File: e2e/native/php/test_consumer.php
<?php
declare(strict_types=1);

require __DIR__ . '/vendor/autoload.php';
foreach (glob(__DIR__ . '/Generated/*.php') ?: [] as $file) require_once $file;

use Brick\Math\BigInteger;
use NativeBindings\EdgeCases;
use NativeBindings\EdgeCasesComplexStruct;
use NativeBindings\NativeCases;
use NativeBindings\NativeCasesGrid;
use NativeBindings\NamingCases;
use NativeBindings\Token;
use NativeBindings\TokenClient;
use NativeBindings\TokenTransactionOptions;

function check(bool $condition, string $message): void
{
    if (!$condition) throw new RuntimeException($message);
}

function rejects(callable $body, string $message): void
{
    try {
        $body();
    } catch (InvalidArgumentException $error) {
        return;
    }
    throw new RuntimeException($message);
}

function word(string $hex): string
{
    return str_pad($hex, 64, '0', STR_PAD_LEFT);
}

function offline(): void
{
    $spender = '0x0000000000000000000000000000000000000002';
    $amount = BigInteger::of('340282366920938463463374607431768211456');
    $calldata = Token::encodeApprove($spender, $amount);
    check($calldata === '0x095ea7b3' . word('2') . word('100000000000000000000000000000000'), 'big integer calldata');
    check(Token::decodeBalanceOfResult('0x' . word('100000000000000000000000000000000'))->isEqualTo($amount), 'big integer result');
    rejects(static fn() => Token::decodeBalanceOfResult('0x' . word('2') . word('0')), 'trailing result accepted');
    rejects(static fn() => Token::decodeMintResult('0x' . word('0')), 'nonempty void result accepted');
    check(Token::decodeInvalidRecipientError(Token::INVALID_RECIPIENT_ERROR_SELECTOR) instanceof \NativeBindings\TokenInvalidRecipientError, 'zero-arg error');
    rejects(static fn() => Token::decodeInvalidRecipientError(Token::INVALID_RECIPIENT_ERROR_SELECTOR . word('0')), 'trailing error data accepted');

    $input = new EdgeCasesComplexStruct(BigInteger::of(7), $spender, '0x' . str_repeat('11', 32), true, 'hello', BigInteger::of(9));
    $tuple = EdgeCases::encodeComplexStructValue($input);
    $decoded = EdgeCases::decodeComplexStructValue($tuple);
    check($decoded->id->isEqualTo(BigInteger::of(7)) && $decoded->name === 'hello' && $decoded->active, 'tuple roundtrip');
    check(str_starts_with(EdgeCases::encodeProcessComplex($input), EdgeCases::PROCESS_COMPLEX_SELECTOR), 'tuple calldata');
    $nested = EdgeCases::decodeNestedArrayResult('0x' . word('20') . word('1') . word('20') . word('2') . word('1') . word('2'));
    check(count($nested) === 1 && $nested[0][1]->isEqualTo(BigInteger::of(2)), 'nested array result');
    $error = EdgeCases::decodeUnauthorizedError(EdgeCases::UNAUTHORIZED_ERROR_SELECTOR);
    check($error instanceof \NativeBindings\EdgeCasesUnauthorizedError, 'custom error');
    check(EdgeCases::decodeInvalidInputError(EdgeCases::UNAUTHORIZED_ERROR_SELECTOR) === null, 'error selector mismatch');
    $payload = word('40') . word('80') . word('3') . str_pad(bin2hex('bad'), 64, '0') . word('1') . str_pad('01', 64, '0');
    $typedError = EdgeCases::decodeInvalidInputError(EdgeCases::INVALID_INPUT_ERROR_SELECTOR . $payload);
    check($typedError !== null && $typedError->reason === 'bad' && $typedError->data === '0x01', 'payload custom error');

    $grid = new NativeCasesGrid([[BigInteger::of(1), BigInteger::of(2)]], [true, false], ['0x' . str_repeat('ab', 32)]);
    $decodedGrid = NativeCases::decodeGridValue(NativeCases::encodeGridValue($grid));
    check($decodedGrid->rows[0][1]->isEqualTo(BigInteger::of(2)) && $decodedGrid->flags === [true, false], 'nested tuple arrays');
    check(substr(NamingCases::encodeFooBar(true), 0, 10) !== substr(NamingCases::encodeFooBar2(BigInteger::of(1)), 0, 10), 'overload selectors');

    $topic = '0x' . str_repeat('ab', 32);
    $filter = NativeCases::filterIndexedReferencesEvent($spender, '0x1', '0x2', $topic);
    check($filter['topics'][1] === $topic && $filter['topics'][2] === null, 'indexed reference hash filter');
    $decodedLog = NativeCases::decodeIndexedReferencesEvent([
        'topics' => [NativeCases::INDEXED_REFERENCES_EVENT_TOPIC, $topic, $topic, $topic],
        'data' => '0x',
    ]);
    check($decodedLog !== null && $decodedLog->label === $topic && $decodedLog->values === $topic, 'indexed reference hash decoding');
    rejects(static fn() => NativeCases::encodeGridValue(new NativeCasesGrid([[BigInteger::of(1), BigInteger::of(2)]], ['false', false], ['0x' . str_repeat('ab', 32)])), 'nested bool accepted string');
}

function rpc(): void
{
    $url = getenv('ATG_RPC_URL');
    if ($url === false || $url === '') return;
    $from = '0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266';
    $address = (string)getenv('ATG_TOKEN_ADDRESS');
    $client = new TokenClient($address, $url, (string)getenv('ATG_PRIVATE_KEY'), $from, (int)getenv('ATG_CHAIN_ID'));
    $options = new TokenTransactionOptions($client->gasPrice(), BigInteger::of(500000));
    $amount = BigInteger::of('340282366920938463463374607431768211456');
    $mintHash = Token::sendMint($client, $from, $amount, $options);
    $mint = receipt($client, $mintHash);
    check(Token::callBalanceOf($client, $from)->isEqualTo($amount), 'typed mint read');
    $mintEvent = Token::decodeTransferEvent($mint['logs'][0]);
    check($mintEvent !== null && $mintEvent->amount->isEqualTo($amount), 'mint event decode');

    $approveHash = Token::sendApprove($client, $from, BigInteger::of(17), $options);
    $approval = receipt($client, $approveHash);
    check(Token::callAllowance($client, $from, $from)->isEqualTo(BigInteger::of(17)), 'typed allowance');
    $approvalEvent = Token::decodeApprovalEvent($approval['logs'][0]);
    check($approvalEvent !== null && $approvalEvent->amount->isEqualTo(BigInteger::of(17)), 'approval event decode');
    $filter = Token::filterApprovalEvent($address, $approval['blockNumber'], $approval['blockNumber'], $from, $from);
    check(count($client->getLogs($filter)) === 1, 'indexed event query');
    rejects(static fn() => Token::sendApprove($client, $from, BigInteger::of(1), new TokenTransactionOptions($client->gasPrice(), BigInteger::of(500000), BigInteger::of(1))), 'nonpayable accepted value');
    try {
        $client->call(Token::encodeTransfer('0x0000000000000000000000000000000000000000', BigInteger::of(1)));
    } catch (RuntimeException $error) {
        check(str_contains($error->getMessage(), 'JSON-RPC error'), 'revert surfaced');
        return;
    }
    throw new RuntimeException('expected transfer revert');
}

function receipt(TokenClient $client, string $hash): array
{
    $deadline = microtime(true) + 20;
    do {
        $receipt = $client->getTransactionReceipt($hash);
        if ($receipt !== null) {
            check(hexdec($receipt['status']) === 1, 'transaction reverted');
            return $receipt;
        }
        usleep(100000);
    } while (microtime(true) < $deadline);
    throw new RuntimeException('receipt timeout');
}

offline();
rpc();
echo "PHP generated consumer passed\n";

File: e2e/native/python/test_consumer.py
"""Executable offline and disposable-Anvil tests for generated web3.py bindings."""
import os

from eth_abi import encode as abi_encode
from eth_utils import keccak
from web3 import Web3
from web3.exceptions import ContractLogicError

from Generated.Token import TokenContract
from Generated.EdgeCases import EdgeCasesContract
from Generated.NamingCases import NamingCasesContract


def check(condition: bool, message: str) -> None:
    if not condition:
        raise AssertionError(message)


def offline() -> None:
    w3 = Web3()
    address = Web3.to_checksum_address("0x0000000000000000000000000000000000000001")
    spender = Web3.to_checksum_address("0x0000000000000000000000000000000000000002")
    token = TokenContract(address, w3)
    calldata = token.encode_approve(spender, 42)
    check(calldata.startswith("0x095ea7b3") and len(calldata) == 138, "typed calldata")
    check(token.decode_balance_of_result("0x" + abi_encode(["uint256"], [42]).hex()) == 42, "typed output")
    for data in ("0x" + abi_encode(["uint256"], [42]).hex() + "00" * 32, "0x00"):
        try:
            if len(data) > 10:
                token.decode_balance_of_result(data)
            else:
                token.decode_mint_result(data)
        except ValueError:
            pass
        else:
            raise AssertionError("malformed output accepted")
    check(token.decode_invalid_recipient_error("0x9c8d2cd2") == (), "custom error")
    try:
        token.approve(spender, 42, {"value": 1})
    except ValueError as error:
        check("not payable" in str(error), "nonpayable guard")
    else:
        raise AssertionError("nonpayable value accepted")
    edge = EdgeCasesContract(address, w3)
    check(edge.encode_process_complex((42, spender, bytes(32), True, "tuple", 7)).startswith("0x"), "tuple encoding")
    nested = abi_encode(["uint256[][]"], [[[1, 2]]])
    check(edge.decode_nested_array_result("0x" + nested.hex()) == ((1, 2),), "nested array decoding")
    error = keccak(text="InvalidInput(string,bytes)")[:4] + abi_encode(["string", "bytes"], ["bad", b"\x01"])
    check(edge.decode_invalid_input_error("0x" + error.hex()) == ("bad", b"\x01"), "typed custom error")
    naming = NamingCasesContract(address, w3)
    check(naming.encode_foo_bar(True)[:10] != naming.encode_foo_bar_2(1)[:10], "distinct canonical overloads")


def rpc() -> None:
    url = os.getenv("ATG_RPC_URL")
    if not url:
        return
    w3 = Web3(Web3.HTTPProvider(url))
    chain_id = int(os.environ["ATG_CHAIN_ID"])
    check(w3.eth.chain_id == chain_id, "chain ID")
    account = w3.eth.account.from_key(os.environ["ATG_PRIVATE_KEY"])
    address = Web3.to_checksum_address(os.environ["ATG_TOKEN_ADDRESS"])
    token = TokenContract(address, w3)
    amount = 1 << 128

    def submit(tx: dict) -> dict:
        signed = account.sign_transaction(tx)
        hash_ = w3.eth.send_raw_transaction(signed.raw_transaction)
        receipt = w3.eth.wait_for_transaction_receipt(hash_, timeout=20)
        check(receipt.status == 1, "transaction reverted")
        return receipt

    base = {"from": account.address, "chainId": chain_id, "gas": 500_000, "gasPrice": w3.eth.gas_price}
    mint_tx = token.mint(account.address, amount, dict(base, nonce=w3.eth.get_transaction_count(account.address)))
    mint_receipt = submit(mint_tx)
    check(token.balance_of(account.address) == amount, "typed read after mint")
    check(len(mint_receipt.logs) == 1, "missing mint event")
    transfer = token.decode_transfer_log(mint_receipt.logs[0])
    check(transfer["args"]["to"] == account.address and transfer["args"]["amount"] == amount, "typed transfer decode")
    filtered = token.filter_transfer(from_block=mint_receipt.blockNumber, to_block=mint_receipt.blockNumber)
    check(len(filtered) == 1, "historical filter")

    approve_tx = token.approve(account.address, 17, dict(base, nonce=w3.eth.get_transaction_count(account.address)))
    approve_receipt = submit(approve_tx)
    check(token.allowance(account.address, account.address) == 17, "typed allowance")
    check(token.decode_approval_log(approve_receipt.logs[0])["args"]["amount"] == 17, "approval event")

    zero = Web3.to_checksum_address("0x0000000000000000000000000000000000000000")
    try:
        w3.eth.call({"to": address, "from": account.address, "data": token.encode_transfer(zero, 1)})
    except (ContractLogicError, ValueError):
        pass
    else:
        raise AssertionError("expected transfer revert")


if __name__ == "__main__":
    offline()
    rpc()
    print("Python generated web3.py consumer passed")
